server/src/main/java/org/apache/iotdb/db/sync/sender/transfer/DataTransferManager.java

```diff
@@ -0,0 +1,659 @@
+      for (String sgName : allSG) {
+        if (!storageGroups.isEmpty() && !storageGroups.contains(sgName)) {
+          continue;
+        }
+        lastLocalFilesMap.putIfAbsent(sgName, new HashSet<>());
+        syncLog = new SyncSenderLogger(getSyncLogFile());
+        try {
+          ResultStatus status = serviceClient.init(sgName);
+          if (!status.success) {
+            throw new SyncConnectionException("Unable init receiver because " + status.errorMsg);
+          }
+        } catch (TException | SyncConnectionException e) {
+          throw new SyncConnectionException("Unable to connect to receiver", e);
+        }
+        logger.info("Sync process starts to transfer data of storage group {}", sgName);
+        syncDeletedFilesNameInOneGroup(sgName,
+            deletedFilesMap.getOrDefault(sgName, new HashSet<>()));
+        syncDataFilesInOneGroup(sgName, toBeSyncedFilesMap.getOrDefault(sgName, new HashSet<>()));
+      }
+
+    } catch (SyncConnectionException e) {
+      logger.error("cannot finish sync process", e);
+    } finally {
+      if (syncLog != null) {
+        syncLog.close();
+      }
+      syncStatus = false;
+    }
+  }
+
+  @Override
+  public void syncDeletedFilesNameInOneGroup(String sgName, Set<File> deletedFilesName)
+      throws IOException {
+    if (deletedFilesName.isEmpty()) {
+      logger.info("There has no deleted files to be synced in storage group {}", sgName);
+      return;
+    }
+    syncLog.startSyncDeletedFilesName();
+    logger.info("Start to sync names of deleted files in storage group {}", sgName);
+    for (File file : deletedFilesName) {
+      try {
+        if (serviceClient.syncDeletedFileName(file.getName()).success) {
+          logger.info("Receiver has received deleted file name {} successfully.", file.getName());
+          lastLocalFilesMap.get(sgName).remove(file);
+          syncLog.finishSyncDeletedFileName(file);
+        }
+      } catch (TException e) {
+        logger.error("Can not sync deleted file name {}, skip it.", file);
+      }
+    }
+    logger.info("Finish to sync names of deleted files in storage group {}", sgName);
+  }
+
+  @Override
+  public void syncDataFilesInOneGroup(String sgName, Set<File> toBeSyncFiles)
+      throws SyncConnectionException, IOException {
+    if (toBeSyncFiles.isEmpty()) {
+      logger.info("There has no new tsfiles to be synced in storage group {}", sgName);
+      return;
+    }
+    syncLog.startSyncTsFiles();
+    logger.info("Sync process starts to transfer data of storage group {}", sgName);
+    int cnt = 0;
+    for (File tsfile : toBeSyncFiles) {
+      cnt++;
+      try {
+        File snapshotFile = makeFileSnapshot(tsfile);
+        // firstly sync .restore file, then sync tsfile
+        syncSingleFile(new File(snapshotFile.getAbsolutePath() + TsFileResource.RESOURCE_SUFFIX));
+        syncSingleFile(snapshotFile);
+        lastLocalFilesMap.get(sgName).add(tsfile);
```

Comment:
In the design doc, you said that if a tsfile is not sealed, we shouldn't transfer it. But this is not present in the code, maybe I can't find it. Could you give some explanation about this?

Reply:
Please see more details on line 105-118 in  the method `getCurrentLocalFiles()` of class `SyncFileManager`

Reply:
I find it!  Awesome!